Support nested C# exceptions

eadred opened the pull request:
* Inner exceptions
* Exceptions thrown from asynchronous code

igorpeshansky commented:
@eadred Thank you for the contribution!

This plugin is designed to work with [Stackdriver Error Reporting](https://cloud.google.com/error-reporting/docs/), which should recognize the resulting multiline message as a stack trace. I'm verifying this with the team right now whether the backend would recognize a nested C# exception if the stack trace were joined into a single log entry. If the backend won't recognize these, there's no point in handling them in the plugin, and we should file an Error Reporting feature request.

igorpeshansky commented:
I've confirmed with the Error Reporting team that all 3 new exceptions were picked up by Error Reporting, but maybe not in the form you'd expect. For example, for the `CSHARP_AGGREGATE_EXC` value from the test, only the content between the arrows was captured.

For the record, a self-service test can be done by running:
```sh
echo "System.AggregateException: One or more errors occurred. (This is an exception) ---> System.InvalidOperationException: This is an exception
  at ExampleApp.AsyncExceptionExample.LowestLevelMethod() in c:/ExampleApp/ExampleApp/AsyncExceptionExample.cs:line 28
  at ExampleApp.AsyncExceptionExample.ThirdLevelMethod() in c:/ExampleApp/ExampleApp/AsyncExceptionExample.cs:line 23
  at ExampleApp.AsyncExceptionExample.SecondLevelMethod() in c:/ExampleApp/ExampleApp/AsyncExceptionExample.cs:line 17
  at ExampleApp.AsyncExceptionExample.TopLevelMethod() in c:/ExampleApp/ExampleApp/AsyncExceptionExample.cs:line 12
  --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
  at System.Threading.Tasks.Task.Wait()
  at ExampleApp.Program.Main(String[] args) in c:/ExampleApp/ExampleApp/Program.cs:line 12
---> (Inner Exception #0) System.InvalidOperationException: This is an exception
  at ExampleApp.AsyncExceptionExample.LowestLevelMethod() in c:/ExampleApp/ExampleApp/AsyncExceptionExample.cs:line 28
  at ExampleApp.AsyncExceptionExample.ThirdLevelMethod() in c:/ExampleApp/ExampleApp/AsyncExceptionExample.cs:line 23
  at ExampleApp.AsyncExceptionExample.SecondLevelMethod() in c:/ExampleApp/ExampleApp/AsyncExceptionExample.cs:line 17
  at ExampleApp.AsyncExceptionExample.TopLevelMethod() in c:/ExampleApp/ExampleApp/AsyncExceptionExample.cs:line 12<---
" | \
gcloud beta error-reporting events report \
  --service manual \
  --service-version test-version \
  --project=$(gcloud config get-value project) \
  --message-file=/dev/stdin
```
(see the [gcloud docs](https://cloud.google.com/sdk/gcloud/reference/beta/error-reporting/events/report)).

If the ingestion results don't look ok, let's open a feature request against Error Reporting before proceeding with this PR. For the values that work correctly, we can start reviewing those portions of the PR.

igorpeshansky commented on lib/fluent/plugin/exception_detector.rb:
```diff
@@ -59,6 +59,22 @@ def self.supported
+
+      # C# aggregate exception
+      rule([:java_after_exception, :java],
+            /^---> \(Inner Exception #[0-9]+\)/,
```
These look sufficiently different from Java that they merit a separate set of rules for C# (possibly sharing some of the Java rules programmatically, e.g., by copying the rules and doing symbol substitution). @qingling128, WDYT?

qingling128 replied:
Agreed. At least we should separate the names (not having `java_after_exception` in a `C#` exception.

eadred replied:
The problem I'm having implementing this is there's not a good way of distinguishing between what is a C# exception and what is a Java exception when first transitioning to a `*_start_exception` state. Essentially both have to look for `Exception` in the text.

Because of this, if the `JAVA_RULES` are added first to `ALL_RULES` before `CSHARP_RULES`, then C# exceptions go off down the Java state machine path, and therefore never encounter C# specific rules like the ones I'm trying to implement. Likewise, if `CSHARP_RULES` were to appear before `JAVA_RULES` then Java exceptions get picked up by the C# state machine.

igorpeshansky commented on test/plugin/test_exception_detector.rb:
```diff
@@ -256,6 +256,50 @@ class ExceptionDetectorTest < Test::Unit::TestCase
   at System.Threading.Thread.StartInternal () [0x00000] in <filename unknown>:0
 END
 
+  CSHARP_NESTED_EXC = <<END.freeze
```
It might be time to define an array of exceptions per language. This is a much larger refactoring than we should reasonably expect you to do in this PR, but if you're game, that's great. If not, we could try to do it ourselves, and then rebase your PR on top of the result. @qingling128, WDYT?

qingling128 replied:
I'm fine with merging this PR once it's approved and refactoring them into arrays per language later. This way we unblock the use case earlier.

eadred commented on lib/fluent/plugin/exception_detector.rb:
```diff
@@ -59,6 +59,23 @@ def self.supported
+      # C# exception from async code
+      rule([:java_after_exception, :java],
+           %r{^---\sEnd\sof\sstack\strace\sfrom\sprevious\s
+           location\swhere\sexception\swas\sthrown\s---}x,
```
The linter doesn't like this ('Use // around regular expression.'), but without it I end up with a line that is too long for the linter. Have you got any suggestions about how to split the regex across multiple lines (Ruby isn't my language)?

I could perhaps define a '+' operator for Regexp, as suggested [here](https://www.ruby-forum.com/t/concatenation-of-regular-expressions/87885/4).

igorpeshansky replied:
Try the `/x` modifier, as suggested [here](https://stackoverflow.com/questions/3762183/split-ruby-regex-over-multiple-lines).

eadred replied:
Thanks. Done.

eadred commented:
Sorry about the force push. I've now restored the original commit as it was at the time of the first review.

I've now managed to get the C# rules into a separate set of rules (in the most recent commit https://github.com/GoogleCloudPlatform/fluent-plugin-detect-exceptions/commit/a991faaa09c74bb19a40bb6cb9a377bc021441ea). As I mentioned in an earlier comment, I've now had to cope with the fact that it is ambiguous as far as the rules are concerned whether you are seeing a C# or a Java exception. To implement this I've had to support the state machine having more than one possible current state. For example, it might be in either the `:java_start_exception` or the `:csharp_start_exception` state when it see a line like _"foo.bar.baz.Exception: Something went wrong."_

Unfortunately these changes have made the bench tests around 3 to 4 times slower because the _ExceptionDetector.transition_ method can't exit early once it has found a match.

eadred commented:
I checked how the new C# exceptions looked in error reporting, and it seems it is only the aggregate exception that doesn't get picked up properly. It is actually the other two types of exception I am more interested in getting handled properly, so if it meant this PR could get merged faster I could drop the logic for handling the aggregate exception. What do you think?

qingling128 commented:
@eadred - Thanks for refactoring the code. Now I see why it was challenging to separate `csharp` and `java` states. We care a lot of performance, so 3X - 4X processing time is sufficient to make an exception for this case. We probably want to revert the https://github.com/GoogleCloudPlatform/fluent-plugin-detect-exceptions/commit/a991faaa09c74bb19a40bb6cb9a377bc021441ea commit.

As to the `aggregate` exception, let's drop that part (please open a feature request with Error Reporting backend if needed) so that we can get this PR in.

eadred commented:
Thanks for that.

I've now reverted https://github.com/GoogleCloudPlatform/fluent-plugin-detect-exceptions/commit/a991faaa09c74bb19a40bb6cb9a377bc021441ea (as https://github.com/GoogleCloudPlatform/fluent-plugin-detect-exceptions/pull/58/commits/7fa27c0d842d804a7fd81724a2916389fcaacca5) and also removed the handling for aggregate exceptions (https://github.com/GoogleCloudPlatform/fluent-plugin-detect-exceptions/pull/58/commits/58889d353f3f7e92866d053dda31000f0353a1b2).

I've raised an issue with the Error Reporting team for handling AggregateExceptions - https://issuetracker.google.com/issues/139292720.

PS. I realised that my handling of aggregate exceptions wasn't dealing with the case where there are multiple inner exceptions, so I did put in a fix for that (https://github.com/GoogleCloudPlatform/fluent-plugin-detect-exceptions/pull/58/commits/6ea0cb7f6d935668782073d3bd4770c24c3ddcc2). Although this has all been reverted anyway, I may want to re-add these changes in a future PR.

qingling128 commented:
Thanks @eadred! This looks good to me. Seems like there are still pending comments from @igorpeshansky. He is on vacation and will be back on Monday (Aug 19). Is this something that can wait or is it urgent?

eadred commented:
Thanks @qingling128. No problem if this is delayed - it isn't urgent.

qingling128 commented:
Cool, thanks. Will wait for @igorpeshansky to take a second look once he's back.

eadred commented on lib/fluent/plugin/exception_detector.rb:
```diff
@@ -59,6 +59,18 @@ def self.supported
       rule(:java_after_exception, /^[\r\n]*$/, :java_after_exception),
       rule([:java_after_exception, :java], /^[\t ]+(?:eval )?at /, :java),
+
+      # C# nested exception
```
Done

eadred commented on lib/fluent/plugin/exception_detector.rb:
```diff
@@ -59,6 +59,18 @@ def self.supported
+
+      # C# exception from async code
+      rule([:java_after_exception, :java],
+           /^---\sEnd\sof\sstack\strace\sfrom\sprevious\s
```
Done as per the former suggestion.

eadred commented on lib/fluent/plugin/exception_detector.rb:
```diff
@@ -59,6 +59,18 @@ def self.supported
+
+      # C# nested exception
+      rule([:java_after_exception, :java],
+           /^[\t ]+--- End of inner exception stack trace ---/,
```
Added

eadred commented on lib/fluent/plugin/exception_detector.rb:
```diff
@@ -59,6 +59,18 @@ def self.supported
+      # C# exception from async code
+      rule([:java_after_exception, :java],
+           /^---\sEnd\sof\sstack\strace\sfrom\sprevious\s
+           location\swhere\sexception\swas\sthrown\s---/x,
```
Added

xflag commented:
Could you release a new version for this issue?
Now I need this enhancement,I use .net core running in docker swarm,then collect log from docker logging driver to fluentd which add this plugin for multiline error logs.

qingling128 commented:
I'll work with our oncall @davidbtucker to do a release.

qingling128 commented:
Sent https://github.com/GoogleCloudPlatform/fluent-plugin-detect-exceptions/pull/65.

qingling128 commented:
Done: https://github.com/GoogleCloudPlatform/fluent-plugin-detect-exceptions/releases/tag/0.0.13
